uc-crux-llvm/src/UCCrux/LLVM/Equivalence.hs

```diff
@@ -0,0 +1,261 @@
+import qualified UCCrux.LLVM.Config as Config
+import           UCCrux.LLVM.Config (UCCruxLLVMOptions)
+import           UCCrux.LLVM.Context.App (AppContext, log)
+import           UCCrux.LLVM.Context.Module (ModuleContext)
+import           UCCrux.LLVM.Logging (Verbosity(Low))
+import           UCCrux.LLVM.Run.Simulate (UCCruxSimulationResult, explanations)
+import           UCCrux.LLVM.Run.Result (SomeBugfindingResult(..))
+import           UCCrux.LLVM.Run.Loop (zipResults)
+{- ORMOLU_ENABLE -}
+
+data CrashDiff = CrashDiff
+  { _bugDiff :: Set (Located TruePositive),
+    -- | Failed assertions
+    _assertionDiff :: Set (Located ()),
+    -- | Unclassified errors
+    _unclassifiedDiff :: Set (Located ()),
+    _unfixedDiff :: Set (Located Unfixed),
+    _unfixableDiff :: Set (Located Unfixable)
+  }
+  deriving (Eq, Ord)
+
+instance Semigroup CrashDiff where
+  CrashDiff s1 s2 s3 s4 s5 <> CrashDiff t1 t2 t3 t4 t5 =
+    CrashDiff (s1 <> t1) (s2 <> t2) (s3 <> t3) (s4 <> t4) (s5 <> t5)
+
+instance Monoid CrashDiff where
+  mempty = CrashDiff Set.empty Set.empty Set.empty Set.empty Set.empty
+
+newtype NonEmptyCrashDiff = NonEmptyCrashDiff CrashDiff
+  deriving (Eq, Ord, Monoid, Semigroup)
+
+crashDiffNonEmpty :: CrashDiff -> Maybe NonEmptyCrashDiff
+crashDiffNonEmpty diff =
+  if diff == mempty
+    then Nothing
+    else Just (NonEmptyCrashDiff diff)
+
+crashDiffIsEmpty :: CrashDiff -> Bool
+crashDiffIsEmpty = isNothing . crashDiffNonEmpty
+
+ppCrashDiff :: CrashDiff -> Doc Void
+ppCrashDiff (CrashDiff posDiff assertDiff unclassDiff unfixedDiff unfixableDiff) =
+  PP.vcat $
+    catMaybes
+      [ printDiff posDiff "Likely bugs at",
+        printDiff assertDiff "Assertion failures at",
+        printDiff unclassDiff "Unclassified errors at",
+        printDiff unfixedDiff "Unfixed errors at",
+        printDiff unfixableDiff "Unfixable errors at"
+      ]
+  where
+    headerWithBullets header lst =
+      PP.nest 2 (PP.vcat (header : map ((PP.pretty ("-" :: Text) PP.<+>) . PP.pretty) lst))
+    printDiff :: Set (Located a) -> Text -> Maybe (Doc Void)
+    printDiff diff header =
+      if Set.null diff
+        then Nothing
+        else
+          Just $
+            headerWithBullets
+              (PP.pretty header)
+              (map (show . What4.plSourceLoc . location) (toList diff))
+
+renderCrashDiff :: CrashDiff -> Text
+renderCrashDiff =
+  PP.renderStrict . PP.layoutPretty PP.defaultLayoutOptions . ppCrashDiff
+
+-- | Compute the \"crash difference\" between two sets of results.
+--
+-- This ignores:
+--
+-- * timeouts,
+-- * errors that are missing annotations, and
+-- * executions that encountered unimplemented features of UC-Crux-LLVM.
+--
+-- The behavior of this function is documented in the README, updates should be
+-- reflected there.
+--
+-- The composition of this function with 'crashDiffIsEmpty' is a partial order
```

Comment:
Did this mean `crashDiffNonEmpty`? `crashDiffIsEmpty` returns `Bool`

Reply:
Yeah, this is a bit vague - "composition" doesn't fully make sense here. I was thinking that a (total) function `[Located (Explanation ...)] -> [Located (Explanation ...)] -> Bool` is semantically comparable to a subset of the cross product `([Located (Explanation ...)], [Located (Explanation ...)])`, i.e. a binary relation (and in particular, a partial order) on `[Located (Explanation ...)]`. I can add more of that explanation inline here.